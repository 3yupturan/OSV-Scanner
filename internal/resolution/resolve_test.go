package resolution_test

import (
	"cmp"
	"context"
	"slices"
	"testing"

	"deps.dev/util/resolve"
	"deps.dev/util/resolve/dep"
	"github.com/google/osv-scanner/internal/resolution"
	"github.com/google/osv-scanner/internal/resolution/clienttest"
	"github.com/google/osv-scanner/internal/resolution/manifest"
	"github.com/google/osv-scanner/internal/testutility"
)

func checkResult(t *testing.T, result *resolution.ResolutionResult) {
	t.Helper()
	snap := testutility.NewSnapshot()
	snap.MatchText(t, result.Graph.String())

	type minimalVuln struct {
		ID               string
		DevOnly          bool
		ProblemChains    [][]resolve.Edge
		NonProblemChains [][]resolve.Edge
	}

	minVulns := make([]minimalVuln, len(result.Vulns))
	for i, v := range result.Vulns {
		minVulns[i] = minimalVuln{
			ID:               v.Vulnerability.ID,
			DevOnly:          v.DevOnly,
			ProblemChains:    make([][]resolve.Edge, len(v.ProblemChains)),
			NonProblemChains: make([][]resolve.Edge, len(v.NonProblemChains)),
		}
		for j, c := range v.ProblemChains {
			minVulns[i].ProblemChains[j] = c.Edges
		}
		for j, c := range v.NonProblemChains {
			minVulns[i].NonProblemChains[j] = c.Edges
		}
	}
	slices.SortFunc(minVulns, func(a, b minimalVuln) int {
		return cmp.Compare(a.ID, b.ID)
	})
	snap.MatchJSON(t, minVulns)
}

func aliasType(knownAs string) dep.Type {
	typ := dep.NewType()
	typ.AddAttr(dep.KnownAs, knownAs)
	return typ
}

func TestResolve(t *testing.T) {
	type requirement struct {
		name    string
		version string
		typ     dep.Type
		groups  []string
	}
	tests := []struct {
		name         string
		version      string
		system       resolve.System
		universe     string
		requirements []requirement
	}{
		{
			name:     "test",
			version:  "1.0.0",
			system:   resolve.NPM,
			universe: "./fixtures/universe.yaml",
			requirements: []requirement{
				{
					name:    "bob",
					version: "^1.0.0",
				},
				{
					name:    "charlie",
					version: "2.0.1",
				},
				{
					name:    "charlie",
					version: "2.0.0",
					typ:     aliasType("sharlie"),
				},
				{
					name:    "alice",
					version: "1.0.0",
					typ:     aliasType("zalice"),
				},
				{
					name:    "alice",
					version: "^1.2.3",
					groups:  []string{"dev"},
				},
			},
		},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			cl := clienttest.NewMockResolutionClient(t, tt.universe)
			var m manifest.Manifest
			m.Root = resolve.Version{
				VersionKey: resolve.VersionKey{
					PackageKey: resolve.PackageKey{
						Name:   tt.name,
						System: tt.system,
					},
					Version:     tt.version,
					VersionType: resolve.Concrete,
				},
			}
			m.Groups = make(map[resolve.PackageKey][]string)
			m.Requirements = make([]resolve.RequirementVersion, len(tt.requirements))
			for i, req := range tt.requirements {
				pk := resolve.PackageKey{
					Name:   req.name,
					System: tt.system,
				}
				m.Requirements[i] = resolve.RequirementVersion{
					VersionKey: resolve.VersionKey{
						PackageKey:  pk,
						Version:     req.version,
						VersionType: resolve.Requirement,
					},
					Type: req.typ,
				}
				m.Groups[pk] = req.groups
			}

			res, err := resolution.Resolve(context.Background(), cl, m)
			if err != nil {
				t.Fatalf("error resolving: %v", err)
			}
			checkResult(t, res)
		})
	}
	t.Fail()
}
