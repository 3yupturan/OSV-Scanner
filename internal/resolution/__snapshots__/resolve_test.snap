
[TestResolve/test - 1]
test 1.0.0
├─ 1: reg|Selector="" | bob@^1.0.0 1.2.3
├─ reg|Selector="" | charlie@2.0.1 2.0.1
│  └─ reg|Selector="" | alice@~1.2.0 1.2.3
│     └─ $1@^1.2.3
├─ reg|KnownAs="sharlie"|Selector="" | charlie@2.0.0 2.0.0
│  └─ reg|Selector="" | alice@~1.2.0 1.2.3
│     └─ $1@^1.2.3
├─ reg|KnownAs="zalice"|Selector="" | alice@1.0.0 1.0.0
└─ reg|Selector="" | alice@^1.2.3 1.3.0
   └─ $1@*

---

[TestResolve/test - 2]
[
  {
    "ID": "OSV-123-456",
    "DevOnly": false,
    "ProblemChains": [
      [
        {
          "From": 0,
          "To": 4,
          "Requirement": "1.0.0",
          "Type": {}
        }
      ],
      [
        {
          "From": 2,
          "To": 6,
          "Requirement": "~1.2.0",
          "Type": {}
        },
        {
          "From": 0,
          "To": 2,
          "Requirement": "2.0.1",
          "Type": {}
        }
      ],
      [
        {
          "From": 3,
          "To": 7,
          "Requirement": "~1.2.0",
          "Type": {}
        },
        {
          "From": 0,
          "To": 3,
          "Requirement": "2.0.0",
          "Type": {}
        }
      ]
    ],
    "NonProblemChains": []
  },
  {
    "ID": "OSV-456-789",
    "DevOnly": false,
    "ProblemChains": [
      [
        {
          "From": 0,
          "To": 1,
          "Requirement": "^1.0.0",
          "Type": {}
        }
      ],
      [
        {
          "From": 6,
          "To": 1,
          "Requirement": "^1.2.3",
          "Type": {}
        },
        {
          "From": 2,
          "To": 6,
          "Requirement": "~1.2.0",
          "Type": {}
        },
        {
          "From": 0,
          "To": 2,
          "Requirement": "2.0.1",
          "Type": {}
        }
      ],
      [
        {
          "From": 7,
          "To": 1,
          "Requirement": "^1.2.3",
          "Type": {}
        },
        {
          "From": 3,
          "To": 7,
          "Requirement": "~1.2.0",
          "Type": {}
        },
        {
          "From": 0,
          "To": 3,
          "Requirement": "2.0.0",
          "Type": {}
        }
      ]
    ],
    "NonProblemChains": [
      [
        {
          "From": 5,
          "To": 1,
          "Requirement": "*",
          "Type": {}
        },
        {
          "From": 0,
          "To": 5,
          "Requirement": "^1.2.3",
          "Type": {}
        }
      ]
    ]
  }
]
---
